tf_helpers/tf_dot_general.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""
+Construct an equivalent general dot operation as that in JAX -
+    <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+
+Although there is an implementation in TF XLA, avoid directly using XLA when
+possible.
+
+Zhibo Zhang, 2020.06.30
+"""
+
+import tensorflow as tf
+from tensorflow.python.ops import numpy_ops as tf_np
+import string
+
+
+def _minus(a, b):
+  return [x for x in a if x not in b]
+
+
+def compose_output_rep(lhs_rep, rhs_rep, lhs_contraction, rhs_contraction,
+                        lhs_batch, rhs_batch):
+  """ Compose the output string representation.
+
+  e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+        aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+
+  Args:
+    lhs_rep: A string representation for the left-hand side input array
+    rhs_rep: A string representation for the right-hand side input array
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+    lhs_batch: Sequence[int] (the batch dimensions of lhs)
+    rhs_batch: Sequence[int] (the batch dimensions of rhs)
+
+  Returns:
+    A string representation of the result array.
+  """
+  output_rep = []
+  for dim in lhs_batch:
+    output_rep.append(lhs_rep[dim])
+
+  for i in _minus(range(len(lhs_rep)), lhs_batch + lhs_contraction):
+    output_rep.append(lhs_rep[i])
+  for i in _minus(range(len(rhs_rep)), rhs_batch + rhs_contraction):
+    output_rep.append(rhs_rep[i])
+  return ''.join(output_rep)
+
+
+def non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction):
+  """ Compute the non-batched matrix multiplication.
+
+  If it is the general non-batched/single-batched matrix multiplication,
+  use the highly optimized kernel `tf.tensordot` to handle it.
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+
+  Returns:
+    An array that contains the result.
+  """
+  return tf.tensordot(lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction)))
+
+
+def tf_dot_general(lhs, rhs, dimension_numbers):
```

Comment:
Thanks again for this awesome effort to make NT work with TF!

Initial question: if we're trying to mirror the necessary lax functionality in TF, for the purpose of easier import / invocation swapping, should this method

1) be in a file called `lax.py` like this one https://github.com/DarrenZhang01/Neural_Tangents_TensorFlow/blob/master/tf_jax_stax/tf_lax.py
2) be called `dot_general`, similarly, for the purpose of matching `lax.py`?
3) Have the `precision` argument, to fully match the signature (even if it goes unused - not sure if TF has a straightforward analog).

Reply:
Yes, great insight! Thanks, Roman! I will make the changes.

Reply:
By the way, do you think I should add `reduce_window`, `dot_general`, `conv_general_dilated` all together into `lax` and make one single Pull Request or I should do them one by one? I am not sure which is more convenient for you. Thanks, Roman @romanngg !

Reply:
I would suggest submitting all `lax.py` file together with all the methods that you need there + also the respective tests!

Reply:
Sounds good!